From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pewpew <pewpew@nyaa.cat>
Date: Sun, 18 Jan 2026 15:10:00 +0900
Subject: [PATCH] Optimize block entity ticker removals


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 579bbba4e823d4d0318e58759ca732b7c8e4d865..b382133d5602cfa6192e43993447af09f6c7aed9 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -112,7 +112,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         .add(new ExplosionParticleInfo(ParticleTypes.POOF, 0.5F, 1.0F))
         .add(new ExplosionParticleInfo(ParticleTypes.SMOKE, 1.0F, 1.0F))
         .build();
-    public final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList();
+    public final io.papermc.paper.pewpew.BlockEntityTickersList blockEntityTickers = new io.papermc.paper.pewpew.BlockEntityTickersList(); // Pewpew - optimize block entity removals
     protected final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
     private boolean tickingBlockEntities;
@@ -1401,13 +1401,11 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         boolean runsNormally = this.tickRateManager().runsNormally();
 
         int tickedEntities = 0; // Paper - rewrite chunk system
-        var toRemove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<TickingBlockEntity>(); // Paper - Fix MC-117075; use removeAll
-        toRemove.add(null); // Paper - Fix MC-117075
         for (this.tileTickPosition = 0; this.tileTickPosition < this.blockEntityTickers.size(); this.tileTickPosition++) { // Paper - Disable tick limiters
             TickingBlockEntity tickingBlockEntity = this.blockEntityTickers.get(this.tileTickPosition);
             // Spigot end
             if (tickingBlockEntity.isRemoved()) {
-                toRemove.add(tickingBlockEntity); // Paper - Fix MC-117075; use removeAll
+                this.blockEntityTickers.markAsRemoved(this.tileTickPosition); // Pewpew - optimize block entity removals
             } else if (runsNormally && this.shouldTickBlocksAt(tickingBlockEntity.getPos())) {
                 tickingBlockEntity.tick();
                 // Paper start - rewrite chunk system
@@ -1417,7 +1415,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
                 // Paper end - rewrite chunk system
             }
         }
-        this.blockEntityTickers.removeAll(toRemove); // Paper - Fix MC-117075
+        this.blockEntityTickers.removeMarkedEntries(); // Pewpew - optimize block entity removals
 
         this.tickingBlockEntities = false;
         this.spigotConfig.currentPrimedTnt = 0; // Spigot
