From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pewpew <pewpew@nyaa.cat>
Date: Sun, 18 Jan 2026 15:10:00 +0900
Subject: [PATCH] Optimize block entity ticker removals


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 56659181dc817ea028428eabf5937308c04b8cc5..fe783f0a53f7287b3629dcf2c97baad7b9a2e886 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -118,7 +118,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         .add(new ExplosionParticleInfo(ParticleTypes.POOF, 0.5F, 1.0F))
         .add(new ExplosionParticleInfo(ParticleTypes.SMOKE, 1.0F, 1.0F))
         .build();
-    public final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList();
+    public final io.papermc.paper.pewpew.BlockEntityTickersList blockEntityTickers = new io.papermc.paper.pewpew.BlockEntityTickersList(); // Pewpew - optimize block entity removals
     protected final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
     private boolean tickingBlockEntities;
@@ -1490,14 +1490,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         boolean tickBlockEntities = this.tickRateManager().runsNormally();
 
         int tickedEntities = 0; // Paper - rewrite chunk system
-        // Paper start - Fix MC-117075 use removeAll
-        final it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<@Nullable TickingBlockEntity> toRemove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
-        toRemove.add(null);
         for (int tickerIndex = 0; tickerIndex < this.blockEntityTickers.size(); tickerIndex++) {
             final TickingBlockEntity ticker = this.blockEntityTickers.get(tickerIndex);
-        // Paper end - Fix MC-117075 use removeAll
             if (ticker.isRemoved()) {
-                toRemove.add(ticker); // Paper - Fix MC-117075 use removeAll
+                this.blockEntityTickers.markAsRemoved(tickerIndex); // Pewpew - optimize block entity removals
             } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
                 ticker.tick();
                 // Paper start - rewrite chunk system
@@ -1508,7 +1504,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
         }
 
-        this.blockEntityTickers.removeAll(toRemove); // Paper - Fix MC-117075 use removeAll
+        this.blockEntityTickers.removeMarkedEntries(); // Pewpew - optimize block entity removals
         this.tickingBlockEntities = false;
     }
 
