From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pewpew <pewpew@nyaa.cat>
Date: Sun, 18 Jan 2026 15:12:00 +0900
Subject: [PATCH] Skip ServerEntity delta check when unchanged


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 3a817c92147bb78c60238f3ddc1644b721aed7a2..b34574a5cd31bacdc8c24c678c9e84c65d9127e5 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -197,6 +197,7 @@ public class ServerEntity {
 
                 if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) {
                     Vec3 deltaMovement = this.entity.getDeltaMovement();
+                    if (deltaMovement != this.lastSentMovement) { // Pewpew - skip distance check when unchanged
                     double d = deltaMovement.distanceToSqr(this.lastSentMovement);
                     if (d > 1.0E-7 || d > 0.0 && deltaMovement.lengthSqr() == 0.0) {
                         this.lastSentMovement = deltaMovement;
@@ -214,6 +215,7 @@ public class ServerEntity {
                             this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
                         }
                     }
+                    }
                 }
 
                 if (packet != null) {
