From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pewpew <pewpew@nyaa.cat>
Date: Sun, 18 Jan 2026 15:15:00 +0900
Subject: [PATCH] Track per-world MSPT


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index f142cf80d2458c36e492143db4cb551955ed767b..643f75193f58854adfc935523da9639d335174ed 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1852,7 +1852,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                     worldProfiler.push("tick");
 
                     try {
+                        long start = Util.getNanos(); // Pewpew - track per-world MSPT
                         level.tick(haveTime);
+                        long elapsed = Util.getNanos() - start; // Pewpew - track per-world MSPT
+                        level.tickTimes5s.add(this.tickCount, elapsed);
+                        level.tickTimes10s.add(this.tickCount, elapsed);
+                        level.tickTimes1m.add(this.tickCount, elapsed);
                     } catch (Throwable t) {
                         CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
                         level.fillReportDetails(report);
@@ -1884,7 +1889,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 profiler.push("tick");
 
                 try {
+                    long start = Util.getNanos(); // Pewpew - track per-world MSPT
                     level.tick(haveTime);
+                    long elapsed = Util.getNanos() - start; // Pewpew - track per-world MSPT
+                    level.tickTimes5s.add(this.tickCount, elapsed);
+                    level.tickTimes10s.add(this.tickCount, elapsed);
+                    level.tickTimes1m.add(this.tickCount, elapsed);
                 } catch (Throwable t) {
                     CrashReport report = CrashReport.forThrowable(t, "Exception ticking world");
                     level.fillReportDetails(report);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 229ff740e0418a05d824a67401ae6f105b4c3642..88f4ea38db72289158518978ae17d9ddfc071d25 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -604,6 +604,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
     // Paper end - chunk tick iteration
 
+    // Pewpew start - track per-world MSPT
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes1m = new MinecraftServer.TickTimes(1200);
+    // Pewpew end
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
