From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pewpew <pewpew@nyaa.cat>
Date: Sun, 18 Jan 2026 15:15:00 +0900
Subject: [PATCH] Track per-world MSPT


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 6ad52a430fe3a27844e8ff5e1293380c11bd147d..780c39c43866462430c647e2b9058010a0514283 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1822,7 +1822,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                     worldProfiler.push("tick");
 
                     try {
+                        long start = Util.getNanos(); // Pewpew - track per-world MSPT
                         serverLevel.tick(hasTimeLeft);
+                        long elapsed = Util.getNanos() - start; // Pewpew - track per-world MSPT
+                        serverLevel.tickTimes5s.add(this.tickCount, elapsed);
+                        serverLevel.tickTimes10s.add(this.tickCount, elapsed);
+                        serverLevel.tickTimes1m.add(this.tickCount, elapsed);
                     } catch (Throwable var7) {
                         CrashReport crashReport = CrashReport.forThrowable(var7, "Exception ticking world");
                         serverLevel.fillReportDetails(crashReport);
@@ -1861,7 +1866,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 profilerFiller.push("tick");
 
                 try {
+                    long start = Util.getNanos(); // Pewpew - track per-world MSPT
                     serverLevel.tick(hasTimeLeft);
+                    long elapsed = Util.getNanos() - start; // Pewpew - track per-world MSPT
+                    serverLevel.tickTimes5s.add(this.tickCount, elapsed);
+                    serverLevel.tickTimes10s.add(this.tickCount, elapsed);
+                    serverLevel.tickTimes1m.add(this.tickCount, elapsed);
                 } catch (Throwable var7) {
                     CrashReport crashReport = CrashReport.forThrowable(var7, "Exception ticking world");
                     serverLevel.fillReportDetails(crashReport);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index dc65503a2d785d64d37b76b0303f51cf66d9769a..cf574b16cca17f7ff24583acc88d5666aecf7697 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -598,6 +598,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
     // Paper end - chunk tick iteration
 
+    // Pewpew start - track per-world MSPT
+    public final MinecraftServer.TickTimes tickTimes5s = new MinecraftServer.TickTimes(100);
+    public final MinecraftServer.TickTimes tickTimes10s = new MinecraftServer.TickTimes(200);
+    public final MinecraftServer.TickTimes tickTimes1m = new MinecraftServer.TickTimes(1200);
+    // Pewpew end
+
     public ServerLevel(
         MinecraftServer server,
         Executor dispatcher,
