From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Codex <codex@local>
Date: Sun, 18 Jan 2026 16:39:15 +0000
Subject: [PATCH] Synchronize map id allocation


diff --git a/net/minecraft/world/level/saveddata/maps/MapIndex.java b/net/minecraft/world/level/saveddata/maps/MapIndex.java
index 6b3ce2ded53a7ea46cf7f0cb49043a022f6d9ee7..60d790abd6729dedb3dcba44e56618b3e915bb61 100644
--- a/net/minecraft/world/level/saveddata/maps/MapIndex.java
+++ b/net/minecraft/world/level/saveddata/maps/MapIndex.java
@@ -25,7 +25,7 @@ public class MapIndex extends SavedData {
         this.lastMapId = lastMapId;
     }
 
-    public MapId getNextMapId() {
+    public synchronized MapId getNextMapId() {
         MapId id = new MapId(++this.lastMapId);
         this.setDirty();
         return id;
