From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix Bot <w@c.at>
Date: Mon, 19 Jan 2026 18:34:25 +0000
Subject: [PATCH] Cache spawn biomes by quart position


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index c322ce879795a6d530680568c7dcf499b21a38a4..011e53a4512d709350e1fe77c92ba24eca6cde8d 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -757,15 +757,18 @@ public final class NaturalSpawner {
             if (!this.contains(pos)) {
                 return this.level.getBiome(pos);
             }
-            long key = pos.asLong();
+            int quartX = QuartPos.fromBlock(pos.getX());
+            int quartY = QuartPos.fromBlock(pos.getY());
+            int quartZ = QuartPos.fromBlock(pos.getZ());
+            long key = QuartPos.asLong(quartX, quartY, quartZ);
             Holder<Biome> cached = this.biomeCache.get(key);
             if (cached != null) {
                 return cached;
             }
             Holder<Biome> biome = this.chunk.getNoiseBiome(
-                QuartPos.fromBlock(pos.getX()),
-                QuartPos.fromBlock(pos.getY()),
-                QuartPos.fromBlock(pos.getZ())
+                quartX,
+                quartY,
+                quartZ
             );
             this.biomeCache.put(key, biome);
             return biome;
