From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix Bot <w@c.at>
Date: Mon, 19 Jan 2026 18:50:22 +0000
Subject: [PATCH] Fix quart biome cache key packing


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index aefd696667351fe9d09857e742120193e302b048..5cc98ccdbd409e7c016794e3996c9fc3309bc8c4 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -729,7 +729,7 @@ public final class NaturalSpawner {
             int quartX = QuartPos.fromBlock(pos.getX());
             int quartY = QuartPos.fromBlock(pos.getY());
             int quartZ = QuartPos.fromBlock(pos.getZ());
-            long key = QuartPos.asLong(quartX, quartY, quartZ);
+            long key = BlockPos.asLong(quartX, quartY, quartZ);
             Holder<Biome> cached = this.biomeCache.get(key);
             if (cached != null) {
                 return cached;
